### PR TITLE
Add simple BV solver

### PR DESCRIPTION
**Pull request by mpreiner**
This PR adds a simple BV solver that sends bit-blasting lemmas to the internal MiniSat.

### CODE REVIEW

File: src/smt/set_defaults.cpp [hunk 1]
```diff
@@ -125,6 +125,14 @@ void setDefaults(LogicInfo& logic, bool isInternalSubsolver)
     }
   }
 
+  /* BVSolver::SIMPLE does not natively support int2bv and nat2bv and need to
```
Comment (aniemetz) [inline src/smt/set_defaults.cpp:128]:
```suggestion
  /* BVSolver::SIMPLE does not natively support int2bv and nat2bv, they need to
```

File: src/theory/bv/bv_solver_simple.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  void bbAtom(TNode node) override;
+  Node getBBAtom(TNode atom) const override;
+  void storeBBAtom(TNode atom, Node atom_bb) override;
+  void storeBBTerm(TNode node, const Bits& bits) override;
+  bool hasBBAtom(TNode atom) const override;
+  Node getStoredBBAtom(TNode node);
+
+  void makeVariable(TNode var, Bits& bits) override;
+
+  bool collectModelValues(TheoryModel* m, const std::set<Node>& relevantTerms);
```
Comment (aniemetz) [inline src/theory/bv/bv_solver_simple.cpp:46]: Document.

File: src/theory/bv/bv_solver_simple.cpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+  prop::SatSolver* getSatSolver() override { return nullptr; }
+
+ private:
+  Node getModelFromSatSolver(TNode a, bool fullModel) override;
+
+  TNodeSet d_variables;
```
Comment (aniemetz) [inline src/theory/bv/bv_solver_simple.cpp:53]: Document members.

File: src/theory/bv/theory_bv_rewriter.h [hunk 4]
```diff
@@ -50,8 +50,9 @@ class TheoryBVRewriter : public TheoryRewriter
 
  private:
   static RewriteResponse IdentityRewrite(TNode node, bool prerewrite = false);
-  static RewriteResponse UndefinedRewrite(TNode node, bool prerewrite = false); 
-  
+  static RewriteResponse UndefinedRewrite(TNode node, bool prerewrite = false);
```
Comment (aniemetz) [inline src/theory/bv/theory_bv_rewriter.h:53]: What's this for? What is it used for?

File: src/theory/bv/bv_solver_simple.cpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+#include "theory/bv/bitblast/lazy_bitblaster.h"
+#include "theory/bv/theory_bv.h"
+#include "theory/bv/theory_bv_utils.h"
+#include "theory/theory_model.h"
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+class BBSimple : public TBitblaster<Node>
```
Comment (ajreynol) [inline src/theory/bv/bv_solver_simple.cpp:28]: Can you add some documentation here above this class.

File: src/theory/bv/bv_solver_simple.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+#include "theory/bv/bv_solver.h"
+
+namespace CVC4 {
+
+namespace theory {
+namespace bv {
+
+class BBSimple;
+
+class BVSolverSimple : public BVSolver
```
Comment (ajreynol) [inline src/theory/bv/bv_solver_simple.h:33]: Add documentation (one paragraph or so) describing the main features of the solver.

File: src/theory/bv/bv_solver_simple.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                                  SubstitutionMap& outSubstitutions) override
+  {
+    return Theory::PPAssertStatus::PP_ASSERT_STATUS_UNSOLVED;
+  }
+
+  bool collectModelValues(TheoryModel* m,
+                          const std::set<Node>& termSet) override;
+
+ private:
+  std::unique_ptr<BBSimple> d_bitblaster;
```
Comment (ajreynol) [inline src/theory/bv/bv_solver_simple.h:59]: Document.

File: src/theory/bv/bv_solver_simple.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  /** Caches lemmas added for BV atoms. */
+  std::unordered_set<Node, NodeHashFunction> d_lemmasCache;
+
+  void addBBLemma(TNode fact);
```
Comment (ajreynol) [inline src/theory/bv/bv_solver_simple.h:63]: Document.

**Review by ajreynol — CHANGES_REQUESTED** [review]
Some documentation requests

File: src/theory/bv/theory_bv_rewriter.h (hunk 4 above)
Comment (mpreiner) [reply src/theory/bv/theory_bv_rewriter.h:53]: That's unused and was probably used to initialize `d_rewriteTable` to indicate that a rewrite for a kind was not implemented. I can remove it if you want.

File: src/theory/bv/bv_solver_simple.cpp [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+{
+}
+
+void BVSolverSimple::addBBLemma(TNode fact)
+{
+  if (!d_bitblaster->hasBBAtom(fact))
+  {
+    d_bitblaster->bbAtom(fact);
+  }
+  if (d_lemmasCache.find(fact) == d_lemmasCache.end())
```
Comment (ajreynol) [inline src/theory/bv/bv_solver_simple.cpp:251]: This lemmas cache appears to be unused?  Regardless I don't think you want to use this: lemmas should be cached based on the user context, moreover lemmas are actually cached by default in the call to TheoryInferenceManager::lemma below.

**Review by ajreynol — CHANGES_REQUESTED** [review]
One minor request, you can remove the `d_lemmaCache` field I believe, otherwise this looks good to me.

File: src/theory/bv/bv_solver_simple.cpp (hunk 9 above)
Comment (mpreiner) [reply src/theory/bv/bv_solver_simple.cpp:251]: done

**Review by ajreynol — APPROVED** [review]

**Pull request merged**